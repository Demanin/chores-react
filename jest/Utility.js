import crypto from 'crypto';

const getRandomString = (length = 20) => {
  return crypto.randomBytes(length).toString('hex');
};

const getRandomNumber = (min = 0, max = 9999999) => {
  min = Math.ceil(min);
  max = Math.floor(max);

  return Math.floor(Math.random() * (max - min)) + min;
};

const getRandomBool = () => {
  return !!getRandomNumber(0, 1);
};

export default {
  getRandomString,
  getRandomNumber,
  getRandomBool,
};
