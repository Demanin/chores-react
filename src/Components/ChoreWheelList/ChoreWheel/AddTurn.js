import React, { Component } from 'react';
import { Button, FormGroup, FormControl } from 'react-bootstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

class AddTurn extends Component
{
  constructor(props) {
    super(props);

    this.handleChange = this.handleChange.bind(this);

    this.state = {
      value: ''
    };
  }

  handleChange(event) {
    this.props.onSaveClick(event.target.value)
  }

  render() {
    if (!this.props.choreWheel.allowAddTurn) {
      return (
        <Button
          variant="success"
          onClick={this.props.onEditClick}
        >
         <FontAwesomeIcon icon="plus" /> Add New Turn
        </Button>
      );
    }

    return (
      <div>
        <FormGroup
          controlId={"turn-adder-" + this.props.id}
          size="sm"
        >
          <FormControl
            as="select"
            onChange={this.handleChange}
          >
            <option key="-1" value="-1">Select a user</option>
            {this.props.userList.map((user, index) =>
              <option key={index} value={user.id}>{user.name}</option>
            )}
          </FormControl>
        </FormGroup>
      </div>
    );
  }
}

export default AddTurn;
