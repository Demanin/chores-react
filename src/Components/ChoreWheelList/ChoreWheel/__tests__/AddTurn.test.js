import AddTurn from 'src/Components/ChoreWheelList/ChoreWheel/AddTurn';
import React from 'react';
import Utility from 'jest/Utility';
import { shallow } from 'enzyme';

describe('AddTurn', () => {
  let onSaveClick;
  let onEditClick;

  beforeEach(() => {
    onSaveClick = jest.fn();
    onEditClick = jest.fn();
  })

  test('add turn not allowed', () => {
    const fixture = shallow(<AddTurn
      choreWheel={{}}
      userList={[]}
      {...{ onEditClick, onSaveClick }}
    />);

    expect(fixture).toMatchSnapshot();
  });

  test('edit click called', () => {
    const fixture = shallow(<AddTurn
      choreWheel={{}}
      userList={[]}
      {...{ onEditClick, onSaveClick }}
    />);

    fixture.simulate('click');

    expect(onEditClick).toHaveBeenCalledTimes(1);
  });

  test('empty user list', () => {
    const fixture = shallow(<AddTurn
      choreWheel={{ allowAddTurn: true }}
      userList={[]}
      {...{ onEditClick, onSaveClick }}
    />);

    expect(fixture).toMatchSnapshot();
  });

  test('FormGroup controlId', () => {
    const id = Utility.getRandomNumber();

    const fixture = shallow(<AddTurn
      choreWheel={{ allowAddTurn: true }}
      userList={[]}
      {...{ onEditClick, onSaveClick, id }}
    />);

    expect(fixture.find('FormGroup').prop('controlId')).toEqual('turn-adder-' + id);
  });

  test('onSaveClick called with event target value', () => {
    const value = Utility.getRandomString();

    const fixture = shallow(<AddTurn
      choreWheel={{ allowAddTurn: true }}
      userList={[]}
      {...{ onEditClick, onSaveClick }}
    />);

    fixture.find('FormControl').simulate('change', { target: { value }});

    expect(onSaveClick).toHaveBeenCalledWith(value);
  });

  test('user list added as options', () => {
    const idA = Utility.getRandomNumber();
    const idB = Utility.getRandomNumber();
    const nameA = Utility.getRandomString();
    const nameB = Utility.getRandomString();

    const fixture = shallow(<AddTurn
      choreWheel={{ allowAddTurn: true }}
      userList={[{ id: idA, name: nameA}, { id: idB, name: nameB }]}
      {...{ onEditClick, onSaveClick }}
    />);

    const select = fixture.find('FormControl');

    expect(select.childAt(0).props()).toEqual({ children: 'Select a user', value: '-1' });
    expect(select.childAt(0).key()).toEqual('-1');
    expect(select.childAt(1).props()).toEqual({ children: nameA, value: idA });
    expect(select.childAt(1).key()).toEqual('0');
    expect(select.childAt(2).props()).toEqual({ children: nameB, value: idB });
    expect(select.childAt(2).key()).toEqual('1');
  });
});
