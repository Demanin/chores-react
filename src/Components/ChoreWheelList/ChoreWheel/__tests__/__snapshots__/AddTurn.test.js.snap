// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTurn add turn not allowed 1`] = `ShallowWrapper {}`;

exports[`AddTurn empty user list 1`] = `ShallowWrapper {}`;
